x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -212,6 +213,14 @@ public ApiKeyService(Settings settings, Clock clock, Client client, XPackLicense
             this.apiKeyAuthCache = CacheBuilder.<String, ListenableFuture<CachedApiKeyHashResult>>builder()
                 .setExpireAfterAccess(ttl)
                 .setMaximumWeight(maximumWeight)
+                .removalListener(notification -> {
+                    if (RemovalNotification.RemovalReason.EVICTED == notification.getRemovalReason()) {
+                        if (getApiKeyAuthCache().count() >= maximumWeight) {
+                            logger.trace("API key with ID [{}] was evicted from the authentication cache, " +
+                                    "possibly due to cache size limit", notification.getKey());
```

Comment:
I don't think trace logging is sufficient.
It will allow us to turn on logging so we can determine whether there is a cache problem, but I don't think it's proactive enough.

Can we look at what it would take to have something that could safely be logged at `warning` instead?

Reply:
Thanks, I think this would be useful since trace logging is very busy. I pushed [some changes](https://github.com/elastic/elasticsearch/pull/74388/commits/725dff3656a867bcaf2fce5571051b792982d56f) trying to add warnings based on eviction rate as a heuristic. I considered the followings in the implementation:

* Regular high rate of eviction due to cache size is a symptom of thrashing
* Assuming 70ms for the expensive hashing, a single core can perform less than 15 hashing per second. This is number is used as the threshold since a single core (smallest node) would be saturated by this rate of eviction. That is, if 15 eviction occurs in a second, it is likely a cache thrashing.
* The warning logging should be throttled. Since agents check in every 10 min, I chose to check the stats every 10 min and issue warnings if possible thrashing is identified.
* The warning logging should be as quickly as possible, it can afford some inaccuracy (use of LongAdder instead of AtomicLong and decrease/increase instead of creating new LongAdder)

Please let me know if you are OK with this approach. I tried to keep the implementation as simplistic as possible. But there are still some level of complexities. I'll add tests for this new piece of code once we agree on the implementation.